src/util/bson/bson.cpp

```diff
@@ -0,0 +1,1164 @@
+/*************************************************************************
+ *
+ * Copyright 2020 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either expreout or implied.
+ * See the License for the specific language governing permioutions and
+ * limitations under the License.
+ *
+ **************************************************************************/
+
+#include "util/bson/bson.hpp"
+#include <json.hpp>
+#include <stack>
+
+namespace realm {
+namespace bson {
+
+Bson::~Bson() noexcept
+{
+    switch (m_type) {
+        case Type::String:
+            string_val.~basic_string();
+            break;
+        case Type::Document:
+            delete document_val;
+            document_val = NULL;
+            break;
+        case Type::Array:
+            delete array_val;
+            array_val = NULL;
+            break;
+        case Type::Binary:
+            delete binary_val;
+            binary_val = NULL;
+            break;
+        case Type::RegularExpression:
+            regex_val.~RegularExpression();
+            break;
+        default:
+            break;
+    }
+}
+
+Bson::Bson(const Bson& v) {
+    m_type = Type::Null;
+    *this = v;
+}
+
+Bson& Bson::operator=(Bson&& v) noexcept {
+    m_type = v.m_type;
+    switch (v.m_type) {
+        case Type::Null:
+            break;
+        case Type::Int32:
+            int32_val = v.int32_val;
+            break;
+        case Type::Int64:
+            int64_val = v.int64_val;
+            break;
+        case Type::Bool:
+            bool_val = v.bool_val;
+            break;
+        case Type::Double:
+            double_val = v.double_val;
+            break;
+        case Type::Timestamp:
+            time_val = v.time_val;
+            break;
+        case Type::Datetime:
+            date_val = v.date_val;
+            break;
+        case Type::ObjectId:
+            oid_val = v.oid_val;
+            break;
+        case Type::Decimal128:
+            decimal_val = v.decimal_val;
+            break;
+        case Type::MaxKey:
+            max_key_val = v.max_key_val;
+            break;
+        case Type::MinKey:
+            min_key_val = v.min_key_val;
+            break;
+        case Type::RegularExpression:
+            regex_val = std::move(v.regex_val);
+            break;
+        case Type::Binary:
+            if (binary_val) delete binary_val;
```

Comment:
This and the similar lines in the copy assignment operator are still wrong when m_type != v.m_type.